splash/resources.py

```diff
@@ -112,21 +115,26 @@ def _writeOutput(self, data, request, content_type=None):
                 request.setHeader(name, value)
             return self._writeOutput(data, request, content_type)
 
-        if isinstance(data, (bool, int, long, float, types.NoneType)):
+        if data is None or isinstance(data, (bool, six.integer_types, float)):
             return self._writeOutput(str(data), request, content_type)
 
         if isinstance(data, BinaryCapsule):
             return self._writeOutput(data.data, request, content_type)
 
-        request.setHeader("content-type", content_type)
+        request.setHeader(b"content-type", content_type)
 
         self._logStats(request)
+        if not isinstance(data, bytes):
+            # Twisted expects bytes as response
+            data = data.encode('utf-8')
         request.write(data)
 
     def _logStats(self, request):
         stats = {
-            "path": request.path,
-            "args": request.args,
+            # Anything we retrieve from Twisted request object contains bytes.
+            # We have to convert it to unicode first for json.dump to succeed.
+            "path": request.path.decode('utf-8'),
+            "args": bytes_to_unicode(request.args),
```

Comment:
This is the only place `bytes_to_unicode` is used, and we don't need to do a recursive decoding here. I think we should drop `bytes_to_unicode` and do decoding the same way as in `RequestOptions.fromrequest`. You can add a (much smaller) utility function to convert Twisted request.args to unicode or just duplicate the code.